core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java

```diff
@@ -71,38 +88,38 @@ public void notifyTransactionDataReceived(int topologyId) {
          log.tracef("Signalling transaction data received for topology %d", topologyId);
       }
       transactionDataTopologyId = topologyId;
+
       transactionDataLock.lock();
+      CompletableFuture<Void> oldFuture = transactionDataFuture;
       try {
-         transactionDataCondition.signalAll();
+         transactionDataFuture = new CompletableFuture<>();
       } finally {
          transactionDataLock.unlock();
+         remoteExecutor.execute(() -> oldFuture.complete(null));
```

Comment:
I don't think it's safe, at least here it wouldn't be. If you don't check topologyId & return the future atomically, you might return future that belongs to a topology with higher id than is the expected one.

One lock or two, little difference.

Using the remote executor rather in STI makes sense.